libslide/src/partial_evaluator.rs

```diff
@@ -5,28 +5,36 @@ mod types;
 use flatten::flatten_expr;
 pub use types::*;
 
-use crate::evaluator_rules::RuleSet;
+use crate::evaluator_rules::{Rule, RuleSet};
 use crate::grammar::*;
 use crate::utils::{hash, normalize};
 
 use std::collections::HashSet;
 use std::error::Error;
 
-/// Evaluates an expression to as simplified a form as possible.
+/// Evaluates a statement to as simplified a form as possible.
 /// The evaluation may be partial, as some values (like variables) may be unknown.
-pub fn evaluate(expr: Stmt, ctxt: EvaluatorContext) -> Result<InternedExpr, Box<dyn Error>> {
+pub fn evaluate(expr: Stmt, ctxt: &EvaluatorContext) -> Result<Stmt, Box<dyn Error>> {
     let mut rule_set = RuleSet::default();
-    for rule in ctxt.rule_denylist {
+    for rule in &ctxt.rule_denylist {
         rule_set.remove(rule)
     }
     let built_rules = rule_set.build()?;
 
-    let mut simplified_expr: InternedExpr = match expr {
-        Stmt::Expr(expr) => expr,
-        // TODO: see below
-        _ => todo!("Evaluation currently only handles expressions"),
-    };
+    Ok(match expr {
+        Stmt::Expr(expr) => Stmt::Expr(evaluate_expr(expr, &built_rules, &ctxt)),
+        Stmt::Assignment(Assignment { var, rhs: expr }) => Stmt::Assignment(Assignment {
+            var,
+            rhs: evaluate_expr(expr, &built_rules, &ctxt),
+        }),
+    })
+}
 
+/// Evaluates an expression to as simplified a form as possible.
+/// The evaluation may be partial, as some values (like variables) may be unknown.
+/// The returned expression is [normalized](crate::utils::grammar::normalize).
+fn evaluate_expr(expr: InternedExpr, rules: &Vec<Rule>, ctxt: &EvaluatorContext) -> InternedExpr {
```

Comment:
We should probably pass a slice instead of a &Vec. https://www.reddit.com/r/rust/comments/8kujd1/newbie_correct_way_to_pass_vector_of_references/

Reply:
good point, done